File: solutions/024-swap-nodes-in-pairs/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func sliceToList(nums []int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range nums {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return dummy.Next
}

func listToSlice(head *ListNode) []int {
	out := []int{}
	for head != nil {
		out = append(out, head.Val)
		head = head.Next
	}
	return out
}

func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{0, head}
	prev := dummy
	current := head
	for current != nil && current.Next != nil {
		next := current.Next
		current.Next = next.Next
		next.Next = current
		prev.Next = next
		prev = current
		current = current.Next
	}
	return dummy.Next
}

func main() {
	head := sliceToList([]int{1, 2, 3, 4})
	got := listToSlice(swapPairs(head))
	fmt.Printf("PASS | %v\n", got)
}

File: solutions/024-swap-nodes-in-pairs/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swap_pairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev = dummy
        current = head
        while current is not None and current.next is not None:
            nxt = current.next
            current.next = nxt.next
            nxt.next = current
            prev.next = nxt
            prev = current
            current = current.next
        return dummy.next


if __name__ == "__main__":
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "_helpers"))
    from testutil import build_list, list_to_slice
    sol = Solution()
    head = build_list([1, 2, 3, 4])
    got = list_to_slice(sol.swap_pairs(head))
    status = "PASS" if got == [2, 1, 4, 3] else "FAIL"
    print(f"{status} | => {got} (expected [2, 1, 4, 3])")
